ext/opentelemetry-ext-datadog/src/opentelemetry/ext/datadog/propagator.py

```diff
@@ -0,0 +1,110 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import typing
+
+from opentelemetry import propagators, trace
+from opentelemetry.context import Context
+from opentelemetry.trace.propagation import (
+    get_span_from_context,
+    set_span_in_context,
+)
+from opentelemetry.trace.propagation.httptextformat import (
+    Getter,
+    HTTPTextFormat,
+    HTTPTextFormatT,
+    Setter,
+)
+
+from .constants import DD_ORIGIN
+
+
+class DatadogFormat(HTTPTextFormat):
+    """Propagator for the Datadog HTTP header format.
+    """
+
+    TRACE_ID_KEY = "x-datadog-trace-id"
+    PARENT_ID_KEY = "x-datadog-parent-id"
+    SAMPLING_PRIORITY_KEY = "x-datadog-sampling-priority"
+    ORIGIN_KEY = "x-datadog-origin"
+
+    def extract(
+        self,
+        get_from_carrier: Getter[HTTPTextFormatT],
+        carrier: HTTPTextFormatT,
+        context: typing.Optional[Context] = None,
+    ) -> Context:
+        trace_id = extract_first_element(
+            get_from_carrier(carrier, self.TRACE_ID_KEY)
+        )
+
+        span_id = extract_first_element(
+            get_from_carrier(carrier, self.PARENT_ID_KEY)
+        )
+
+        origin = extract_first_element(
+            get_from_carrier(carrier, self.ORIGIN_KEY)
+        )
+
+        # TODO: add sampling
+
+        if trace_id is None or span_id is None:
+            return set_span_in_context(trace.INVALID_SPAN, context)
+
+        span_context = trace.SpanContext(
+            trace_id=int(trace_id),
+            span_id=int(span_id),
+            is_remote=True,
+            trace_state=trace.TraceState({DD_ORIGIN: origin}),
+        )
+
+        return set_span_in_context(trace.DefaultSpan(span_context), context)
+
+    def inject(
+        self,
+        set_in_carrier: Setter[HTTPTextFormatT],
+        carrier: HTTPTextFormatT,
+        context: typing.Optional[Context] = None,
+    ) -> None:
+        span = get_span_from_context(context=context)
+        set_in_carrier(
+            carrier, self.TRACE_ID_KEY, format_trace_id(span.context.trace_id),
+        )
+        set_in_carrier(
+            carrier, self.PARENT_ID_KEY, format_span_id(span.context.span_id)
+        )
+        if DD_ORIGIN in span.context.trace_state:
+            set_in_carrier(
+                carrier, self.ORIGIN_KEY, span.context.trace_state[DD_ORIGIN]
+            )
+
+
+def format_trace_id(trace_id: int) -> str:
+    """Format the trace id according to b3 specification."""
+    return str(trace_id & 0xFFFFFFFFFFFFFFFF)
+
+
+def format_span_id(span_id: int) -> str:
+    """Format the span id according to b3 specification."""
+    return str(span_id)
+
+
+def extract_first_element(
```

Comment:
is this a utility method available in the API? I know it's good to not couple too heavily, but this feels like boilerplate for propagators.

Reply:
I see it in `opentelemetry.correlationcontext.propagation` as well as the b3 code. Could be pulled out as a generally applicable http util method. Suggestions on where that might be? Could do this refactoring in a follow-up PR.

Reply:
correlationcontext is interesting. I don't have a strong opinion, maybe opentelemetry.propagation.utils?